### ISSUE TITLE
What about a disconnect function?

### ISSUE THREAD

**Issue by landabaso** [reactjs/react-redux#396]
Is it possible with the current API to disconnect and reconnect a component at the user convenience?
Would a disconnect function be helpful?

Let me show a case where it would help:

In an SPA I keep several "Pages" cached. That means they are still mounted, but hidden.

By doing that, it is very quick to re-display Pages that have been already visited since only a few updates are necessary. And the virtual dom does its job!

Some of these Pages consist of large lists of redux-connected Items. These list Items are not rendered when not visible (because of shouldComponentUpdate). But after a few Pages are cached I end up having thousands of Items and the App gets slower.

The problem is that all Item wrappers (connect containers) are still being called for each item on every action. I end up having thousands of calls to each wrapped component. As I said no extra render happens but these thousands of calls make the app slow.

It would be nice if I could disconnect certain Item components (the ones that are not visible) so that they are not even called somehow. And to reconnect them later when the Page is going to be displayed again (I used to unmount / remount the whole but re-rendering was worse).

I tried to find a way using redux-batched-subscribe but I believe that lib serves a different purpose.

Are there any plans for a disconnect function or any way to achieve what I proposed?

Thanks!

**Comment by gaearon** [comment]
Generally my recommendation would be to just (un)mount them rather than keeping them in memory. If this is not feasible, I suggest you to not use `connect()` there, and instead manually grab the `store` from context, and call `store.subscribe()` and the `unsubscribe` function in returns whenever you see fit.

Even as a feature request, this doesn’t belong to the Redux repo because Redux store API already lets you unsubscribe at any point of time. So I’ll close this as non-actionable here.

**Comment by landabaso** [comment]
I get the point @gaearon.

I had been playing with the idea of exposing a way to disconnect.

I got the idea while making some tests by doing this from my connected container:

``` 
componentDidMount() {
  //setTimeout 0, so that it runs on the next execution context (after parent has mouned too)
  setTimeout(function(){
    if (this._reactInternalInstance._currentElement._owner._instance.unsubscribe){
      this._reactInternalInstance._currentElement._owner._instance.unsubscribe()
    }   
    }.bind(this), 0); 
  }
```
Of course the code above uses undocumented APIs and there are many things that could go wrong.

Also, there's no way to reconnect later. That's way I thought that it might be a nice addition to expose somehow a (dis/re)connect / (un/re)subscribe methods.

I guess I'll need to create my own connect wrapper or simply call (un)subscribe as you suggested.

Thanks!

**Comment by landabaso** [comment]
In fact, I just learned that you can (re)subscribe by doing:

` this._reactInternalInstance._currentElement._owner._instance.trySubscribe()`

And I guess these are the couple of paired functions that should be called to unsubscribe:

` this._reactInternalInstance._currentElement._owner._instance.tryUnsubscribe()`
` this._reactInternalInstance._currentElement._owner._instance.clearCache()`

As I said, this is undocumented (and some issues may arise). Just wanted to share these ideas in case it also help others.

**Comment by gaearon** [comment]
Yea, this can break any time.
Also sorry about 

>Even as a feature request, this doesn’t belong to the Redux repo because Redux store API already lets you unsubscribe at any point of time. So I’ll close this as non-actionable here.

I just realized you're in the right repo. (Notification overload on my side :-). 

As I said it’s unlikely we add explicit support for it and you're better off using store directly.

**Comment by landabaso** [comment]
Thanks @gaearon!

I decided the approach of (un)subscribe anyway since it also makes the App slow.

I'll explain it in case some other people read this issue in the future:

I worked a bit more on the idea. I passed a prop from the parent though intermediate components to the final subscribed list Items: `pauseSubscriptions = true/false`. Then the list items unsubscribed and resubscribed when it's parent Page was hidden or shown.

The problem is that passing that prop triggers a chain of renders which would not have happened without it. I guess there are ways to directly (un)subscribe the children items not passing props but I have the feeling I was over-engineering the problem.

I'll try different approaches to avoid having thousands of connected components. Maybe connecting only from the parent and passing props to the children. This will make the program a bit verbose (lots of props-passing) but may do the trick!

Thanks again @gaeron for such great tools and support!
